File: strings/11_palindrome_permutationII/palindrome_permutationII.cpp
#include <iostream>
#include <cmath>
#include <stdexcept>
#include <cstdint>
#include <array>
#include <string>
#include <vector>
#include <unordered_set>

using namespace std;


class Solution {
public:
    vector<string> generatePalindromes(string s) {
        if (!checkConstraints(s)) {
            throw std::runtime_error("Constraints violated");
        }
        array<uint8_t, 26> _map{};
        vector<char> st(s.size() / 2);
        if (!canPermutePalindrome(s, _map)) {
            return vector<string>();
        }
        char ch = 0;
        int k = 0;
        for (int i = 0; i < 26; i++) {
            if (_map[i] % 2 == 1) {
                ch = static_cast<char>(i + 'a');
            }
            for (int j = 0; j < _map[i] / 2; j++) {
                st[k++] = static_cast<char>(i + 'a');
            }
        }
        permute(st, 0, ch);
        return vector<string>(_set.begin(), _set.end());
    }

private:
    unordered_set<string> _set;
    bool checkConstraints(const string& s) {
        if (s.size() < 1 || s.size() > 16) return false;
        for (const char& ch : s) {
            int ascii = static_cast<int>(ch);
            if (!(ascii >= static_cast<int>('a') && ascii < static_cast<int>('a') + 26)) return false;
        }
        return true;
    }
    bool canPermutePalindrome(string& s, array<uint8_t, 26>& _map) {
        int odd_cnt = 0;
        for (int i = 0; i < s.size(); i++) {
            _map[s[i] - 'a']++;
            if (_map[s[i] - 'a'] % 2 == 0) {
                odd_cnt--;
            } else odd_cnt++;
        }
        // the number of characters with odd number of occurrences can't exceed 1
        // (1 in case of odd s_length and 0 in case of even s_length).
        return odd_cnt <= 1;
    }
    void permute(vector<char>& st, int l, char ch) {
        if (l == st.size()) {
            _set.insert(string(st.begin(), st.end()) + (ch == 0 ? "" : string(1, ch)) + string(st.rbegin(), st.rend()));
        }
        for (int r = l; r < st.size(); r++) {
            if (st[r] != st[l] || r == l) { 
                // filter duplicates which might arise when swaping two same characters, except the same element chosen twice
                swap(st[l], st[r]);
                permute(st, l + 1, ch);
                swap(st[l], st[r]);
            }
        }
    }
};

File: strings/11_palindrome_permutationII/palindrome_permutationII.py
from typing import List, Set, Union


class Solution:
    def generate_palindromes(self, s: str) -> List[str]:
        if not self._check_constraints(s):
            raise ValueError("Constraints violated")
        self.set: Set[str] = set()
        _map = [0] * 26
        st: List[str] = []
        if not self._can_permute_palindrome(s, _map):
            return []
        ch = None
        for i in range(26):
            if _map[i] % 2 == 1:
                ch = chr(i + ord('a'))
            for j in range(_map[i] // 2):
                st.append(chr(i + ord('a')))
        self._permute(st, 0, ch)
        return list(self.set)

    def _can_permute_palindrome(self, s: str, _map: List[int]) -> bool:
        odd_cnt = 0
        for ch in s:
            _map[ord(ch) - ord('a')] += 1
            if _map[ord(ch) - ord('a')] % 2 == 0:
                odd_cnt -= 1
            else:
                odd_cnt += 1
        # the number of characters with odd number of occurrences can't exceed 1
        # (1 in case of odd s_length and 0 in case of even s_length).
        return odd_cnt <= 1

    def _check_constraints(self, s: str) -> bool:
        if len(s) < 1 or len(s) > 16:
            return False
        for ch in s:
            if not (ord(ch) >= ord('a') and ord(ch) < ord('a') + 26):
                return False
        return True

    def _permute(self, st: List[str], left_id: int, ch: Union[str, None]) -> None:
        if left_id == len(st):
            self.set.add("".join(st) + ("" if ch is None else ch) + "".join(st[::-1]))
        for right_id in range(left_id, len(st)):
            if st[right_id] != st[left_id] or right_id == left_id:
                # filter duplicates which might arise when swaping two same characters, except the same element chosen twice
                st[right_id], st[left_id] = st[left_id], st[right_id]
                self._permute(st, left_id + 1, ch)
                st[right_id], st[left_id] = st[left_id], st[right_id]
